### PR TITLE
Issue373

### PR DESCRIPTION
**Pull request by PlatinumCD**
Libelf and Libdwarf installation packages. All tests should pass.

### CODE REVIEW

File: packages/taucmdr/cf/software/tau_installation.py
Comment (nchaimov) [inline packages/taucmdr/cf/software/tau_installation.py:654]: These lines are indented too far, which is causing an `IndentationError` during installation.

**Review by nchaimov — CHANGES_REQUESTED** [review]
In `tau_installation.py` there's a section that is indented too far.

**Review by nchaimov — APPROVED** [review]
Tested locally and appears to work correctly -- in default configuration libdwarf and libelf and built and TAU is built with them by passing `-dwarf=/path/to/dwarf -elf=/path/to/elf`.

File: packages/taucmdr/cf/software/libelf_installation.py
Comment (zbeekman) [inline packages/taucmdr/cf/software/libelf_installation.py:37]: Need mirror(s) here too. See my other review comment on libdwarf.

@PlatinumCD if these are unchanged from upstream, then it would be great if you could find and add the appropriate download mirrors from the upstream packages. (i.e. a link to download the same version of the library as the one fetched from the uoregon TAU website) If, however, these sources have changes etc. then I'll take care of hosting a mirror on fs.paratools.com/tau-mirror and adding the source repo mirrors here and you don't need to worry about it. Let's see what @nchaimov knows about whether these are changed from upstream.

File: packages/taucmdr/cf/software/tau_installation.py
Comment (zbeekman) [inline packages/taucmdr/cf/software/tau_installation.py:29]: Please do not make style/whitespace changes as part of commits (and ideally PRs, but there is a time and a place for that) that change functionality. Style/whitespace changes should be in their own commit. In this instance it is also changing intentional spacing.

Please add back one newline here.

File: packages/taucmdr/cf/software/tau_installation.py
Comment (zbeekman) [inline packages/taucmdr/cf/software/tau_installation.py:171]: Please revert (add back) this line

File: packages/taucmdr/cf/software/tau_installation.py
Comment (zbeekman) [inline packages/taucmdr/cf/software/tau_installation.py:242]: Please revert (add back) this line

File: packages/taucmdr/cf/software/tau_installation.py
Comment (zbeekman) [inline packages/taucmdr/cf/software/tau_installation.py:464]: no ws changes please

File: packages/taucmdr/cf/software/tau_installation.py
Comment (zbeekman) [inline packages/taucmdr/cf/software/tau_installation.py:486]: here too

File: packages/taucmdr/cf/software/tau_installation.py
Comment (zbeekman) [inline packages/taucmdr/cf/software/tau_installation.py:489]: and here

File: packages/taucmdr/cli/commands/trial/tests/test_create.py
Comment (zbeekman) [inline packages/taucmdr/cli/commands/trial/tests/test_create.py:140]: This test looks good. It would be nice to have tests that verify that libdwarf is not passed to TAU's configure script when `--libdwarf none` and that it is under normal circumstances (on Linux). I'm trying to think of the best place to put such a test, but it's not immediately obvious to me where it should go. I'll think about this some more and then once I have a better idea I can point you in the right direction or implement it myself. This shouldn't block this PR, so I'll put it on my todo list.

**Review by zbeekman — CHANGES_REQUESTED** [review]
Overall this looks great, @PlatinumCD!

There are two things that should definitely be addressed before merging, however:

1. The introduction and deletion of whitespace where there are no other changes in the diff hunk. Addition of whitespace isn't the end of the world, but I've flagged where intentional blank lines have been removed from docstrings. Please revert at least these style/whitespace changes
2. The addition of `-fPIC` to the useropt flags should be removed. I've flagged this.

Also, depending on whether or not we can just use the upstream packages download link for libelf and libdwarf or not, download mirrors should be added. If they are being modified before being hosted on the UO TAU webserver then I (or Nick) can take care of setting up the mirror on fs.paratools.com and modifying the package REPOS. If they are unchanged then it would be great if you could find the corresponding download links for the versions we're using and add the mirrors yourself.

One final note:

It's good to get in the habit of writing more useful commit messages. I certainly not perfect at it, but some good fast resources for learning more about this (and automating it) are:
- https://commit.style/
- https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
- https://thoughtbot.com/blog/5-useful-tips-for-a-better-commit-message

Some good examples of bad commits are at http://whatthecommit.com/

File: packages/taucmdr/cf/software/libdwarf_installation.py
Comment (zbeekman) [reply packages/taucmdr/cf/software/libdwarf_installation.py:38]: It sounds like these sources are unchanged from upstream. If you could find the official download links for matching versions, @PlatinumCD, and add them as mirrors that would be very much appreciated!

File: packages/taucmdr/cf/software/libdwarf_installation.py
Comment (zbeekman) [inline packages/taucmdr/cf/software/libdwarf_installation.py:35]: Also, from the pylint output:
> `W: 35, 0: Unused util imported from taucmdr (unused-import)`

```suggestion
```

File: packages/taucmdr/cf/software/libdwarf_installation.py
Comment (zbeekman) [inline packages/taucmdr/cf/software/libdwarf_installation.py:50]: Also from pylint:
```
C: 50, 0: Wrong continued indentation (add 1 space).
                                                  target_arch, target_os, compilers, REPOS, None, LIBRARIES, HEADERS)
                                                  ^| (bad-continuation)
```

```suggestion
                                                   target_arch, target_os, compilers, REPOS, None, LIBRARIES, HEADERS)
```

File: packages/taucmdr/cf/software/libelf_installation.py
Comment (zbeekman) [inline packages/taucmdr/cf/software/libelf_installation.py:49]: From pylint:
```
************* Module taucmdr.cf.software.libelf_installation
C: 49, 0: Wrong continued indentation (remove 1 space).
                                                  target_arch, target_os, compilers, REPOS, None, LIBRARIES, HEADERS)
                                                 |^ (bad-continuation)
```

```suggestion
                                                 target_arch, target_os, compilers, REPOS, None, LIBRARIES, HEADERS)
```

File: packages/taucmdr/cf/software/tests/test_libdwarf_installation.py
Comment (zbeekman) [inline packages/taucmdr/cf/software/tests/test_libdwarf_installation.py:37]: @nchaimov Should we add a test stub for libelf too?

File: packages/taucmdr/cf/software/tau_installation.py
Comment (zbeekman) [inline packages/taucmdr/cf/software/tau_installation.py:640]: Pylint style:
```
************* Module taucmdr.cf.software.tau_installation
C:640, 0: Wrong continued indentation (remove 1 space).
                                                        (tau_makefile, libdwarf.include_path))
                                                       |^ (bad-continuation)
```

```suggestion
                                                       (tau_makefile, libdwarf.include_path))
```

**Review by zbeekman — CHANGES_REQUESTED** [review]
@PlatinumCD In addition to my previous review, I've gone through and looked at the pylint output too and requested a few style changes (and removal of an unused import). I'll try to make the pylint PR comment more helpful by limiting the scope of it's analysis to just the changed files and lines. That way it will be easier to see what new pylint issues were introduced.

File: packages/taucmdr/cf/software/libdwarf_installation.py
Comment (nchaimov) [inline packages/taucmdr/cf/software/libdwarf_installation.py:3]: Change copyright date of new file to current year
```suggestion
# Copyright (c) 2020, ParaTools, Inc.
```

File: packages/taucmdr/cf/software/libelf_installation.py
Comment (nchaimov) [inline packages/taucmdr/cf/software/libelf_installation.py:3]: Change copyright date of new file to current year
```suggestion
# Copyright (c) 2020, ParaTools, Inc.
```

File: packages/taucmdr/cli/commands/trial/tests/test_create.py
Comment (nchaimov) [reply packages/taucmdr/cli/commands/trial/tests/test_create.py:140]: I think to do this we'll need a new assert function that can examine the output of configuring a package and check for the presence or absence of a string. If libdwarf is used, for example, TAU's configure script will print `Using libdwarf at <path>` and `NOTE: Using DWARF support from libdwarf` while without libdwarf it won't. But there's not currently a convenient way to check the output of configure during a test, as far as I know.

File: packages/taucmdr/cf/software/tests/test_libdwarf_installation.py
Comment (nchaimov) [inline packages/taucmdr/cf/software/tests/test_libdwarf_installation.py:3]: Change copyright date of new file to current year
```suggestion
# Copyright (c) 2020, ParaTools, Inc.
```

**Review by nchaimov — COMMENTED** [review]
Change copyright dates of new files to 2020

File: packages/taucmdr/model/target.py
Comment (nchaimov) [inline packages/taucmdr/model/target.py:436]:
```suggestion
            'default': 'download' if HOST_OS is not DARWIN else None,
            'argparse': {'flags': ('--libelf',),
                         'group': 'software package',
                         'metavar': '(<path>|<url>|download|None)',
                         'action': ParsePackagePathAction},

```

File: packages/taucmdr/cli/commands/trial/tests/test_create.py
Comment (zbeekman) [reply packages/taucmdr/cli/commands/trial/tests/test_create.py:140]: The configuring usually happens when `reset_project_storage()` but you could add a target and switch to it and capture that output more easily. One problem with this approach is that it's not guaranteed to require being reconfigured.

**Review by zbeekman — APPROVED** [review]

**Comment by zbeekman** [comment]
LGTM! Thanks @PlatinumCD!

**Pull request merged**
